python/soft_body_app/soft_application.py

```diff
@@ -0,0 +1,230 @@
+                continue
+            self.usd.set_mesh_positions(body.name, body.x, time)
+    
+    def create_dirichlet_conditions(self, body_name: str, condition_func: Callable):
+        """ Create dirichlet conditions for a body
+
+        Args:
+            body_name (str): The name of the body
+            condition_func (Callable): A function that returns a boolean value, which indicates whether a vertex should be fixed, such as: `lambda x: x[0] > 1.0` would fix vertices with x-coordinates greater than 1.0.
+        """
+        API.create_dirichlet_conditions(self.scene, body_name, condition_func)
+    
+    def create_traction_conditions(self, body_name: str, traction_condition_func: Callable, force: ArrayLike):
+        """ Create traction force for a body
+
+        Args:
+            body_name (str): The name of the body
+            force (ArrayLike): The traction force
+        """
+        API.create_traction_conditions(self.scene, body_name, traction_condition_func, force)
+
+    def simulate(self, T: float = 1.0, render_fps: float = 60.0, update_viewer: bool = True, update_usd: bool = True,non_update_bodies: Sequence[str] = []):
+        """ Simulates the scene for a given duration, `T`, at a specific frame rate, `render_fps`.
+        
+        Args:
+            T (float, optional): The total duration of the simulation in seconds. Defaults to 1.0.
+            render_fps (float, optional): The frames per second at which to render the simulation. Defaults to 60.0.
+            update_viewer (bool, optional): Whether to update the viewer during the simulation. Defaults to True.
+            update_usd (bool, optional): Whether to update the Universal Scene Description (USD) animation during the simulation. Defaults to True.
+            non_update_bodies (Sequence[str], optional): List of body names that shouldn't be updated during the simulation. Defaults to an empty list.
+
+        Raises:
+            ValueError: If USD updates are requested but the USD instance isn't initialized.
+
+        Returns:
+            dict: A log or report of the simulation, detailing various aspects of the scene after the simulation.
+        """
+        sim_dt = self.scene.params.time_step
+        render_dt = 1.0 / render_fps
+        render_steps = int(np.round(T / render_dt))
+        if update_usd:
+            if self.usd is None:
+                raise ValueError("USD is not initialized.")
+            self.usd.set_animation_time(T*render_fps)
+
+        # A nested loop for simulation and rendering, 
+        # outer loop for rendering, inner loop for simulation computation.
+        for step in tqdm(range(render_steps), desc="Rendering Loop:"):
+            dT = render_dt 
+            while dT > 0:
+                dt = min(dT, sim_dt) 
+                API.simulate(self.scene, dt, debug_on=True)
+                dT -= dt
+                if update_viewer:
+                    self.upadte_viewer(non_update_bodies)
```

Comment:
Why is the viewer updated more frequently than the render fps?

Reply:
The `update_viewer` function is used to render the simulation processing here. For this purpose, the `update_viewer` should be executed more frequently than the render fps. In other words, the `update_viewer` is real-time to watch the full simulation processing locally.

Reply:
I find the wording a little confusing, the viewer is actually doing rendering to show the simulation, and saving to a USD file is not really doing any rendering, it is just "baking" the simulation into a file. I guess you can talk about interactive rendering versus offline rendering, but it is still not obvious from the code that there are two types of rendering going on. Perhaps add some inline comments explaining that the viewer is for interactive rendering and tries to keep up with the simulator time-step while rendering is for offline creation of movies and fps control is for the movie creation process?

Reply:
Yeah, no problem. I will fix it.